src/main/java/frc/robot/subsystems/targetvision/TargetVision.kt

```diff
@@ -1,61 +1,203 @@
 
-    data class Measurement(val timestamp: Double, val translation: Translation2d)
+    var robotPoses: MutableList<MutableList<Measurement>> = mutableListOf()
+
+
+    private fun groupConfidence(last: LimelightTarget_Detector, measurement: LimelightTarget_Detector): Double
+    {
+        var score = 0.0
 
-    private fun takeMeasurement(target: LimelightTarget_Retro): Measurement {
-        val targetTranslation = Translation2d(distanceToTarget(target), target.tx)
+        val deltaX = abs(measurement.tx - last.tx)
+        val deltaY = abs(measurement.ty - last.ty)
+        val deltaA = abs(measurement.ta.pow(0.5) - last.ta.pow(0.5))
 
-        return Measurement(timestamp = inputs.lastUpdateTimestamp, translation = targetTranslation)
+        score += (TX_WEIGHT)*( 1 - (deltaX.pow(TX_DROPOFF_RATE) / 54.0.pow(TX_DROPOFF_RATE)) ) //27 = horizontal fov of limelight
+        score += (TY_WEIGHT)*( 1 - (deltaY.pow(TY_DROPOFF_RATE) / 41.0.pow(TY_DROPOFF_RATE)) ) //20.5 = vertical fov of limelight
+        score += (TA_WEIGHT)*( 1 - (deltaA.pow(TA_DROPOFF_RATE) / TA_MAX.pow(TA_DROPOFF_RATE)) )
+
+        return score
     }
 
-    private fun addMeasurement(measurement: Measurement) {
-        measurements.add(measurement)
 
-        if (measurements.size > SAMPLE_NUM) {
-            measurements.removeFirst()
+    fun addMeasurement(measurement: Measurement){
+
+        var grouped = false
+
+        // group targets
+        for (pose in robotPoses) {
+            // TODO: Find good threshold values for grouping
+            if(groupConfidence(pose.last().pose, measurement.pose) > GROUPING_CONFIDENCE_THRESHOLD) {
+                pose.add(measurement)
+                if ((measurement.timestamp - pose.first().timestamp > MAX_SAMPLE_TIME_MS)) { // TODO: find good value
+                    // cleanup old targets
+                    pose.removeFirst()
+                }
+                grouped = true
+                break
+            } else if((measurement.timestamp - pose.last().timestamp > MAX_SAMPLE_TIME_MS)){
+                robotPoses.remove(pose)
+            }
+        }
+
+        if(!grouped){
+            robotPoses.add(mutableListOf(measurement))
         }
+
+
+
+
+
+   }
+
+
+    //give a measurement a rating based on how likely it is to be our current primary target
+    private fun rateTrackingPreference(measurement: List<List<Measurement>>): Int{
+
+        fun getVelocity(samples: Pair<List<Measurement>, Int>): Double{
+            val tVector: List<Double> = samples.first.map{(timestamp, _) -> timestamp}
+            val txVector: List<Double> = samples.first.map{(_, target) -> target.tx}
+
+            // regressing tx
+            return tVector.zip(txVector).sumOf { pair -> pair.first * pair.second } /
+                    tVector.zip(tVector).sumOf { pair -> pair.first * pair.second }
+
+        }
+
+        var measurementIndexed = measurement.map { list -> Pair(list, measurement.indexOf(list))}
```

Comment:
Kotlin's [`withIndex`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/with-index.html) does this for you.

This is also technically wrong in the case of duplicate measurements, and $\mathcal{O}(n^2)$ for no reason.